File: windows/net/samples/enc_mpg_dvd/TranscodeSplit.cs
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

using System;
using PrimoSoftware.AVBlocks;
using System.IO;

namespace EncMpgDvdSample
{
    class OutputStream : System.IO.Stream
    {
        private System.IO.Stream stream_;
        string filename_;

        public OutputStream(string filename)
        {
            filename_ = filename;
            stream_ = new FileStream(filename, FileMode.Create, FileAccess.Write);

        }

        ~OutputStream()
        {
            Dispose(false);
        }

        protected override void Dispose(bool disposing)
        {
            stream_.Dispose();
        }

        public override bool CanRead
        {
            get { return false; }
        }

        public override bool CanSeek
        {
            get { return true; }
        }

        public override bool CanWrite
        {
            get { return true; }
        }

        public override long Length
        {
            get { return stream_.Length; }
        }

        public override long Position
        {
            get { return stream_.Position; }

            set { stream_.Position = value; }
        }

        public override void Flush()
        {
            stream_.Flush();
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            return stream_.Seek(offset, origin);
        }

        public override void SetLength(long value)
        {
            stream_.SetLength(value);
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            throw new NotSupportedException();
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            stream_.Write(buffer, offset, count);
        }
    }

    class Splitter
    {
        public int SplitTime
        {
            set; get;
        }

        public Int64 SplitSize
        {
            set; get;
        }

        public bool ReportProgress
        {
            set; get;
        }

        public bool TranscodeSplit( Options opt, string outFile, string preset, double startTime, 
                                    out double processedTime, out Int64 processedSize, out bool isSplit)
        {
            DeleteFile(outFile);
            processedSize = 0;
            processedTime = 0;
            isSplit = false;

            using (var transcoder = new Transcoder() { AllowDemoMode = true })
            {
                // Input
                var inSocket = new MediaSocket();
                inSocket.File = opt.InputFile;
                inSocket.TimePosition = startTime;
                transcoder.Inputs.Add(inSocket);
                
                using (var outputStream = new OutputStream(outFile))
                {
                    // Output
                    MediaSocket outSocket = MediaSocket.FromPreset(preset);
                    outSocket.Stream = outputStream;
                    transcoder.Outputs.Add(outSocket);

                    outputStream_ = outputStream;
                    transcoder.OnContinue += Transcoder_OnContinue;
                    transcoder.OnProgress += Transcoder_OnProgress;
                    ReportProgress = false;

                    if (opt.SplitTime > 0)
                        SplitTime = opt.SplitTime;

                    if (opt.SplitSize > 0)
                        SplitSize = opt.SplitSize;

                    if (!transcoder.Open())
                    {
                        PrintError("Transcoder open", transcoder.Error);
                        return false;
                    }

                    if (!transcoder.Run())
                    {
                        PrintError("Transcoder run", transcoder.Error);
                        return false;
                    }

                    transcoder.Close();
                    isSplit = isSplit_;
                    processedSize = outputStream.Length;
                    processedTime = processedTime_;
                }
            }

            return true;
        }

        private void Transcoder_OnProgress(object sender, TranscoderProgressEventArgs args)
        {
            processedTime_ = Math.Max(processedTime_, args.CurrentTime);

            if (ReportProgress)
                Console.WriteLine("progress: {0:F1} sec.", args.CurrentTime);
        }

        private void Transcoder_OnContinue(object sender, TranscoderContinueEventArgs args)
        {
            processedTime_ = Math.Max(processedTime_, args.CurrentTime);

            if (outputStream_ != null && SplitSize > 0 && (outputStream_.Length > SplitSize))
            {
                isSplit_ = true;
                args.Continue = false;
                return;
            }

            if (SplitTime > 0.0 && (SplitTime <= args.CurrentTime))
            {
                isSplit_ = true;
                args.Continue = false;
                return;
            }

            args.Continue = true;
        }

        static void DeleteFile(string filename)
        {
            try
            {
                if (System.IO.File.Exists(filename))
                    System.IO.File.Delete(filename);
            }
            catch { }
        }

        static void PrintError(string action, ErrorInfo e)
        {
            if (action != null)
            {
                Console.Write("{0}: ", action);
            }

            if (ErrorFacility.Success == e.Facility)
            {
                Console.WriteLine("Success");
                return;
            }
            else
            {
                Console.WriteLine("{0}, facility:{1} code:{2} hint:{3}", e.Message, (int)e.Facility, e.Code, e.Hint);
            }
        }

        private double processedTime_;
        private OutputStream outputStream_;
        private bool isSplit_;
    }

    class StreamDuration
    {
        public static double GetMinDuration(string file)
        {
            double duration = 0;
            bool  durationInitialized = false;

            using (MediaInfo info = new MediaInfo())
            {
                info.Inputs[0].File = file;

                if (info.Open())
                {
                    foreach(var socket in info.Outputs)
                    {
                        foreach(var pin in socket.Pins)
                        {
                            if(durationInitialized)
                            {
                                duration = Math.Min(duration, pin.StreamInfo.Duration);
                            }
                            else
                            {
                                durationInitialized = true;
                                duration = pin.StreamInfo.Duration;
                            }
                        }
                    }
                }
            }

            return duration;
        }
    }
}

File: macosx/samples/enc_mpg_dvd/TranscodeSplit.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <AVBlocks.h>
#include <PrimoReference++.h>
#include <PrimoUString.h>
#include <string.h>

#include "options.h"
#include "util.h"

using namespace primo::avblocks;

class OutputStream: public primo::Stream
{
    mutable std::ofstream fstream_;
    std::string filename_;
    
public:
    
    OutputStream() {}
    
    ~OutputStream() { close(); }
    
    bool_t setFile(const std::string& filename)
    {
        if (isOpen())
            return true;
        
        filename_ = filename;
        return false;
    }
    
    virtual bool_t open()
    {
        if (isOpen() || filename_.empty())
            return false;
        
        fstream_.open(filename_, std::ios_base::binary);
        return fstream_.is_open();
    }
    
    virtual void close()
    {
        fstream_.close();
    }
    
    virtual bool_t isOpen() const
    {
        return (fstream_.is_open() ? true : false);
    }
    
    virtual bool_t canRead() const { return false; }
    virtual bool_t canWrite() const { return true; }
    virtual bool_t canSeek() const { return true; }
    virtual bool_t read(void* buffer, int32_t bufferSize, int32_t* totalRead) { return false; }
    
    virtual bool_t write(const void* buffer, int32_t dataSize)
    {
        fstream_.write((const char*)buffer, dataSize);
        return fstream_ ? true : false;
    }
    
    virtual int64_t size() const
    {
        return position();
    }
    
    virtual int64_t position() const
    {
        int64_t pos;
        pos = fstream_.tellp();
        
        return pos;
    }
    
    virtual bool_t seek(int64_t position)
    {
        fstream_.seekp(position, std::ios_base::beg);
        return fstream_ ? true : false;
    }
};

class TranscoderSplitCallback: public primo::avblocks::TranscoderCallback
{
public:
    TranscoderSplitCallback():
    outputStream_(NULL),
    splitTime_(0),
    splitSize_(0),
    reportProgress_(true),
    processedTime_(0),
    isSplit_(false)
    { }
    
    void setOutputStream(primo::Stream* outputStream)
    {
        outputStream_ = outputStream;
    }
    
    void setSplitTime(double seconds)
    {
        splitTime_ = seconds;
    }
    
    void setSplitSize(int64_t bytes)
    {
        splitSize_ = bytes;
    }
    
    void setReportProgress(bool reportProgress)
    {
        reportProgress_ = reportProgress;
    }
    
    bool_t onContinue(double currentTime)
    {
        processedTime_ = std::max<double>(processedTime_, currentTime);
        
        if(outputStream_ && splitSize_ &&
           (outputStream_->size() > splitSize_))
        {
            isSplit_ = true;
            return false;
        }
        
        if(splitTime_ && (splitTime_ <= currentTime))
        {
            isSplit_ = true;
            return false;
        }
        
        return true;
    }
    
    void onProgress(double processedTime, double totalTime)
    {
        using namespace std;
        
        processedTime_ = std::max<double>(processedTime_, processedTime);
        
        if (reportProgress_)
        {
            cout.setf(ios::fixed);
            cout.precision(1);
            cout << "progress: " << processedTime << " sec." << endl;
            cout.precision(2);
        }
    }
    
    double processedTime() const { return processedTime_; }
    int64_t processedSize() const { return outputStream_->size(); }
    bool isSplit() const { return isSplit_; }
    
private:
    double processedTime_;
    primo::Stream *outputStream_;
    double splitTime_;
    int64_t splitSize_;
    bool reportProgress_;
    bool isSplit_;
};

bool transcodeSplit(const Options& opt, const std::string& outFile, const char* preset,
                    double startTime, double &processedTime, int64_t &processedSize, bool& isSplit)
{
    deleteFile(outFile.c_str());
    
    primo::ref<Transcoder> transcoder(Library::createTranscoder());
    transcoder->setAllowDemoMode(true);
    
    // Input
    primo::ref<MediaSocket> inSocket (Library::createMediaSocket());
    inSocket->setFile(primo::ustring(opt.in_file));
    inSocket->setTimePosition(startTime);
    transcoder->inputs()->add(inSocket.get());
    
    // Output
    OutputStream outputStream;
    outputStream.setFile(outFile.c_str());
    
    primo::ref<MediaSocket> outSocket(Library::createMediaSocket(preset));
    outSocket->setStream(&outputStream);
    transcoder->outputs()->add(outSocket.get());
    
    TranscoderSplitCallback callback;
    callback.setOutputStream(&outputStream);
    callback.setReportProgress(false);
    
    if (opt.split_time > 0)
        callback.setSplitTime(opt.split_time);
    
    if (opt.split_size > 0)
        callback.setSplitSize(opt.split_size);
    
    transcoder->setCallback(&callback);
    
    if (!transcoder->open())
    {
        printError(transcoder->error());
        return false;
    }
    
    if (!transcoder->run())
    {
        printError(transcoder->error());
        return false;
    }
    
    transcoder->close();
    
    processedTime = callback.processedTime();
    processedSize = callback.processedSize();
    isSplit = callback.isSplit();
    return true;
}

double getMinDuration(const std::string& file)
{
    double minDuration = 0;
    primo::ref<MediaInfo> info ( Library::createMediaInfo() );
    info->setInputFile(primo::ustring(file));
    
    if (info->load())
    {
        minDuration = (info->streams()->count() > 0) ? info->streams()->at(0)->duration() : -1;
        
        for (int i=1; i < info->streams()->count(); ++i)
        {
            minDuration = std::min(minDuration, info->streams()->at(i)->duration());
        }
        return minDuration;
    }
    else
    {
        printError(info->error());
    }
    
    return 0;	
}